experiment/theory.md

```diff
@@ -1 +1,70 @@
+1. The first step is to initialize the distance of all the vertices from the source vertex to infinity. The distance of the source vertex from itself is 0.
+2. Make a list of edges this can be easily done by traversing the adjacency list of the graph.
+3. Now we will relax all the edges. Relaxing an edge means to check if the distance of the destination vertex is greater than the distance of the source vertex plus the weight of the edge. For example the edge we are currently relaxing is (u,v) and distance of u from the source vertex is d\[u\] and distance of v from the source vertex is d\[v\] and weight of the edge is w\[u,v\]. Now we will check if d\[v\] > d\[u\] + w\[u,v\]. If this condition is true then we will update the distance of v from the source vertex to d\[v\] = d\[u\] + w\[u,v\].
+4. We need to repeat the above step V-1 times where V is the number of vertices in the graph.
+
+**Why V-1 times?**
+
+In a graph containing V vertex V-1 is the maximal length of shortest path in a graph. After k iterations of the Bellman–Ford algorithm, you know the minimum distance between any two vertices, when restricted to paths of length at most k. Since this is dynamic programming algorithm so we need to check all the possible combinations, as a result we need to repeat the above step V-1 times.
+
+**Pseudo Code**
+
+    BellmanFord(Graph, source):
+        distance = [] 
+        distance[source] = 0
+        for each vertex v in Graph: 
+            if v != source
+                distance[v] = infinity
+        for i = 1 to size of Graph:
+            for each edge (u, v) with weight w in Graph:
+                if distance[u] + w < distance[v]:
+                    distance[v] = distance[u] + w
+    
+    In the above algorithm distance[v] represents the length of the shortest path from source to v.
+    They are initialized to infinity except for the source vertex, which is initialized to 0.
+    The outer loop is repeated V-1 times, where V is the number of vertices in the graph.
+    The inner loop relaxes all the edges in the graph.
+
+**Time Complexity**
+
+From the above pseudo code we can see that the outer loop is repeated V-1 times and the inner loop is repeated E times where E is the number of edges in the graph. So the time complexity of the algorithm is O(V*E). This is the average time complexity of the algorithm. In the worst case we would have a complete graph then E would be equal to V^2. So the time complexity of the algorithm in the worst case would be O(V^3). We can optimize the algorithm by stooping the outer loop if we have not updated the distance of any vertex in the inner loop. Thus our best case time complxity will be equal to O(E).
```

Comment:
We can optimize the algorithm by stopping the outer loop
in place of
We can optimize the algorithm by stooping the outer loop

Reply:
fixed the spelling mistake